### PR TITLE
GSNGM-349: Topography Profile for polyline

### PR DESCRIPTION
**Pull request by b4l**
TODO:
- [x] Download SVG ~~& PNG~~
- [x] Placement/styling

### CODE REVIEW

**Comment by gberaudo** [comment]
@b4l , it does not look like building.

**Comment by b4l** [comment]
> @b4l , it does not look like building.

Sorry, all fine now after rebasing.

File: src/elements/topographic-profile.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
+import {LitElementI18n} from '../i18n.js';
+import {html} from 'lit';
+import {customElement, state} from 'lit/decorators.js';
+import i18next from 'i18next';
+import draggable from './draggable';
+import ToolboxStore from '../store/toolbox';
+import {cartesianToLv95} from '../projection';
+import * as d3 from 'd3';
```
Comment (gberaudo) [inline src/elements/topographic-profile.ts:8]: I think d3 is not big if you only import the parts you need. Idem in package.json.

File: src/elements/topographic-profile.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+
+  plotProfile() {
+    // set the dimensions and margins of the graph
+    const margin = {top: 10, right: 10, bottom: 40, left: 60};
+    const width = 900 - margin.left - margin.right;
+    const height = 200 - margin.top - margin.bottom;
+
+    // append the svg object
+    const svg = d3.select('#profile-plot')
```
Comment (gberaudo) [inline src/elements/topographic-profile.ts:67]: You should externalize the d3 code to another file which would only depend on d3.

File: src/elements/topographic-profile.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      .append('svg')
+      .attr('width', width + margin.left + margin.right)
+      .attr('height', height + margin.top + margin.bottom)
+      .append('g')
+      .attr('transform', `translate(${margin.left},${margin.top})`);
+
+    d3.json(this.profileServiceUrl('json')).then(data => {
```
Comment (gberaudo) [inline src/elements/topographic-profile.ts:74]: We can keep it tied to Swisstopo profile for now but ideally we would normalize the Swisstopo result to something reasonable (a [[x,y,z,m], [x,y,z,m], ...]).

File: src/elements/topographic-profile.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    super.update(changedProperties);
+  }
+
+  async handleActions(options: GeometryAction) {
+    if (options.action === 'profile') {
+      const geom = ToolboxStore.geometries.getValue().find(geom => geom.id === options.id);
+      this.coords = geom!.positions.map(c => cartesianToLv95(c));
+      this.name = geom!.name;
+      this.hidden = false;
+      await this.updateComplete;
```
Comment (gberaudo) [inline src/elements/topographic-profile.ts:46]: Why do we need to await?

File: src/elements/topographic-profile.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    draggable(this, {
+      allowFrom: '.drag-handle'
+    });
+    super.connectedCallback();
+  }
+
+  update(changedProperties) {
+    if (changedProperties.has('hidden')) {
+      super.hidden = this.hidden;
+    }
```
Comment (gberaudo) [inline src/elements/topographic-profile.ts:36]: What is this for?

**Review by gberaudo — COMMENTED** [review]
Looks like a great start to me.

File: src/elements/topographic-profile.ts (hunk 1 above)
Comment (b4l) [reply src/elements/topographic-profile.ts]: Done in https://github.com/swissgeol/ngm/pull/752/commits/c98999c0583bd034d2635f40584bb7121841a94f

File: src/elements/topographic-profile.ts (hunk 2 above)
Comment (b4l) [reply src/elements/topographic-profile.ts]: Done in https://github.com/swissgeol/ngm/pull/752/commits/c98999c0583bd034d2635f40584bb7121841a94f

File: src/elements/topographic-profile.ts (hunk 3 above)
Comment (b4l) [reply src/elements/topographic-profile.ts]: partially addressed in https://github.com/swissgeol/ngm/pull/752/commits/6c5a34dad5c5e3832d3b10a4c10335258ab1d5f6

File: src/elements/topographic-profile.ts (hunk 4 above)
Comment (b4l) [reply src/elements/topographic-profile.ts:52]: We need to wait for the element to render/update in order for d3 to select the element to draw the graph

File: src/elements/topographic-profile.ts (hunk 5 above)
Comment (b4l) [reply src/elements/topographic-profile.ts:37]: To hide and show the floating window.

**Review by vladyslav-tk — COMMENTED** [review]
Looks good but seems like one point from the description is not implemented: 
 * `Synchronization when the mouse is passing over the profile, the position is shown on the line in the 3D view.`
 
 Is it still actual or we shouldn't implement it?

**Comment by eleu** [comment]
> Is it still actual or we shouldn't implement it?

normally yes, but if we are lacking time let's skip it for the moment.

**Comment by b4l** [comment]
@vladyslav-tk in accordance with @eleu we will address the missing functionality in a follow up issue.

**Review by vladyslav-tk — APPROVED** [review]

**Pull request merged**